PyInstaller/hooks/hook-PyQt5.py

```diff
@@ -18,16 +19,36 @@
                                    'PyQt5')
          if os.path.basename(x[0]) == 'qt.conf']
 
-# Include ICU files, if they exist. See the "Deployment approach" section in
-# ``PyInstaller/utils/hooks/qt.py``.
-[(os.path.join(pyqt5_library_info.location['BinariesPath'], dll),
-  os.path.join('PyQt5', 'Qt', 'bin', dll))
- for dll in ('icudt??.dll', 'icuin??.dll', 'icuuc??.dll')]
-
-# TODO: Include software rendering for OpenGL. See the "Deployment approach". However, because the standard PyQt5 wheel `doesn't include <https://www.riverbankcomputing.com/pipermail/pyqt/2018-June/040387.html>`_ ``d3dcompiler_XX.dll``, this produces failures. When the wheel is updated, this code can be uncommented.
-##binaries = []
-##for dll in ('libEGL.dll', 'libGLESv2.dll', 'd3dcompiler_??.dll', 'opengl32sw.dll'):
-##    dll_path = os.path.join(pyqt5_library_info.location['BinariesPath'], dll)
-##    # Only add files if they exist.
-##    if glob(dll_path):
-##        binaries += [(dll_path, os.path.join('PyQt5', 'Qt', 'bin', dll))]
+
+def include_all_or_none(globs_to_include):
+    """
+    globs_to_include is a list of file name globs
+    Each glob can only match one file.
+    Every file must be available in order for any in the provided list to be
+    included.
+    """
+    to_include = []
+    for dll in globs_to_include:
+        dll_path = os.path.join(pyqt5_library_info.location['BinariesPath'],
+                                dll)
+        dll_file_paths = glob.glob(dll_path)[:1]
+        for dll_file_path in dll_file_paths:
+            file_name = os.path.basename(dll_file_path)
+            dst_dll_path = os.path.join('PyQt5', 'Qt', 'bin', file_name)
+            to_include.append((dll_file_path, dst_dll_path))
+    if len(to_include) == len(globs_to_include):
+        return to_include
+    return []
+
+
+binaries = []
+angle_files = ['libEGL.dll', 'libGLESv2.dll', 'd3dcompiler_??.dll']
```

Comment:
Very nice. I like your use of named variables (`angle_files`) to clarify the code's operation.

Reply:
Would you also add a comment, explaining the need for `include_all_or_none`? That is, that Qt will generate exceptions/crashes if all aren't present, but if none are present it will try other methods of rendering?

Reply:
As far as I know the exception/crash is from PyInstaller checking that all of the dependencies for each files included from `binaries`. https://github.com/pyinstaller/pyinstaller/blob/9a74f7d5522c6d4678b91beae37d611c8adbfd9b/PyInstaller/depend/bindepend.py#L89

Reply:
??? I mean, the `include_all_or_none` function isn't needed for all the wildcarded files -- PyInstaller won't complain about not finding them. I *think* this avoid Qt failures when some but not all files are present, causing an error. Am I totally confused? If not, why do this (except for the `opengl32sw.dll` case, where there's no wildcard present).

Reply:
We want PyInstaller to include `libEGL.dll`, `libGLESv2.dll`, and `d3dcompiler_47.dll`.

Since `d3dcompiler_47.dll` is not available in CI, it will fail.

This is a middle ground, that will work if all the files are included and won't cause a failure during CI.

I think in general it is a good idea to fail when you are packaging something without its dependencies. But we can't have CI fail, so this is required until the PyQt5 Windows wheels include `d3dcompiler_4?.dll`.

Reply:
OK, good -- we're on the same page. Another reason to write it this way is the amazing variety of PyQt5 packages. I don't want PyInstaller to fail due to an incomplete package (such as the current state of the PyQt5 wheel, as you've pointed out), or produce a broken frozen app (such as an app with `libEGL.dll` and `libGLESvs.dll` but not `d3dcompiler_47.dll`.

Anyway, would you add a short comment summarizing this? The idea that these files **must** be included as a group or not included at all, since including them singly produces failures when the frozen app runs? To me, at least, this isn't obvious, hence my request for a comment to capture this (valuable) knowledge.

Reply:
I just used Dependency Walker on both `libEGL.dll` and `libGLESv2.dll` and `d3dcompiler_47.dll` does not show up as a dependency.

I'm not sure why CI fails. Here is an example of the failure. https://ci.appveyor.com/project/matysek/pyinstaller/build/3380/job/wq5j7qnwy4meo2wt#L5314

It fails when including `libEGL.dll`.

```
Failed to load libEGL (Access is denied.)
```

And now this code is not including `libEGL.dll`.

Reply:
As you said, `d3dcompiler_??.dll` is not require to have a working PyQt5 application. As this DLL is not shipped by PyQt5, the hook will never include any DLLs. Thus applications will not work.
Do you think it will be a good idea to remove that DLL from the hook?

Reply:
I don't understand your logic. `d3dcompiler_??.dll` is only included if the other two DLLs are present, guaranteeing a functioning PyQt5. Wheels in the future may include this; self-compiled PyQt5 installs might; perhaps some Linux packages may, so this supports it if present.

Reply:
The logic here is that `d3dcompiler_??.dll` is __not__ required, only the 2 others. A PyQt5 application is working with only `libEGL.dll` and `libGLESv2.dll`.

As Siecje pointed, the `d3dcompiler_??.dll` is not a dependence of any of the other DLLs. Also, it is not shipped with PyQt5. PyQt5 does ship `libEGL.dll` and `libGLESv2.dll`.

My point is the hook will only include DLLs when the 3 are found. But as `d3dcompiler_??.dll` is not found, no DLL are included at all. Could it be possible to include 2 or 3 DLLs, instead of 3 or none?

I think this is working because my application is not doing any use of Qt3D (based on the DLL name), which is the case of most GUI applications. If an application uses Qt3D, then errors logged into the console are good clues for anyone willing to debug.

Or am I missing something?

Reply:
Quoting the [Qt docs](http://doc.qt.io/qt-5/windows-deployment.html#creating-the-application-package), "If dynamic OpenGL is used, you additionally need to include the libraries required for ANGLE and software rendering. For ANGLE, both libEGL.dll and libGLESv2.dll from Qt's lib directory are required as well as the HLSL compiler from DirectX. The HLSL compiler library, d3dcompiler_XX.dll, where XX is the version number that ANGLE (libGLESv2) was linked against."

To me, this means all three are required if any are required. From your comments, I assume that you have a non-openGL PyQt5 app that works with `libEGL.dll` and `libGLESv2.dll`, but doesn't complain if `d3dcompiler_??.dll` is missing, correct? Does it still work if `libEGL.dll` and `libGLESv2.dll` are missing?

On a side note, current PyQt5 wheels do include `d3dcompiler_47.dll`, so this code will include all three. I'm not sure what version of PyQt5 this started with, but it's fairly recent.

Reply:
You are right, if `libGLE*` DLLS are missing, it does not work.
I use PyQt 5.10.2, and the third DLL is missing. Perhaps I just need to include them myself, I did not see the last one is now shipped with PyQt 5.11+.

This is OK like that, thanks for the clarification.